vertical-pod-autoscaler/recommender/metrics/metrics_client.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1lister "k8s.io/kubernetes/pkg/client/listers/core/v1"
+	resourceclient "k8s.io/metrics/pkg/client/clientset_generated/clientset/typed/metrics/v1alpha1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/metrics/pkg/apis/metrics/v1alpha1"
+)
+
+type MetricsClient interface {
+	GetContainersUtilization() ([]*ContainerUtilizationSnapshot, error)
+}
+
+type metricsClient struct {
+	metricsGetter   resourceclient.PodMetricsesGetter
+	podLister       v1lister.PodLister
+	namespaceLister v1lister.NamespaceLister
+}
+
+func NewMetricsClient(metricsGetter resourceclient.PodMetricsesGetter, podLister v1lister.PodLister, namespaceLister v1lister.NamespaceLister) MetricsClient {
+	return &metricsClient{
+		metricsGetter:   metricsGetter,
+		podLister:       podLister,
+		namespaceLister: namespaceLister,
+	}
+}
+
+func (client *metricsClient) GetContainersUtilization() ([]*ContainerUtilizationSnapshot, error) {
+
+	usageSnapshots, err := client.getContainersUsage()
+	containerSpecs, err := client.getContainersSpec()
+	if err != nil {
+		return nil, err
+	}
+
+	utilizationSnapshots, err := mergeIntoUtilization(usageSnapshots, containerSpecs)
+	if err != nil {
+		return nil, err
+	}
+	return utilizationSnapshots, nil
+}
+
+func (client *metricsClient) getContainersSpec() ([]*containerSpecification, error) {
+
+	var containerSpecs []*containerSpecification
+
+	pods, err := client.podLister.List(labels.Everything())
```

Comment:
So you would suggest to cache entire list of containerSpecs and just update those which have changed? 
It will reduce the complexity of this method but entire algorithm will be still O(containers), because we have to iterate through the entire list when calculating utilizations for all containers. 
If you think it is worth it, could you give me an example of how to watch on changes since the last check?